File: npm/src/analysis/surface/splitSentences.js
// @ts-check
/** 종결 부호 뒤 공백에서 문장을 나눈다. 파이썬 analysis/surface/splitSentences.py 와 같다. */

const TERMINAL = /[.?!]+\s+(?=\S)/g;
const ABBREVIATION = /(?:^|[^A-Za-z])[A-Za-z]$/;

/**
 * @typedef {object} Sentence
 * @property {string} text
 * @property {number} start 분석에 넣은 문자열 안의 시작 오프셋
 * @property {number} end
 */

/** @param {string} text @returns {Sentence[]} */
export function splitSentences(text) {
  /** @type {Sentence[]} */
  const sentences = [];
  let start = 0;
  for (const match of text.matchAll(TERMINAL)) {
    const index = /** @type {number} */ (match.index);
    const before = text.slice(0, index);
    if (ABBREVIATION.test(before)) continue;
    const end = index + match[0].trimEnd().length;
    const segment = text.slice(start, end);
    const piece = segment.trim();
    if (piece) sentences.push({ text: piece, start: start + (segment.length - segment.trimStart().length), end });
    start = index + match[0].length;
  }
  const rest = text.slice(start);
  const tail = rest.trim();
  if (tail) {
    const lead = rest.length - rest.trimStart().length;
    sentences.push({ text: tail, start: start + lead, end: start + lead + tail.length });
  }
  return sentences;
}

File: src/hanlint/analysis/splitSentences.py
"""종결 부호 뒤 공백에서 문장을 나눈다. kss 의 punct 모드와 같은 개념이다.

실측: 블로그 글 004 의 222 문장에서 형태소 분석기 (Kiwi) 의 분리와 완전히 일치했고, 기준 말뭉치 390편에서는
쉼표에서 자르거나 경로를 쪼개지 않는 쪽이 이것이었다 (2026-08-29).
나누지 않는 자리는 셋이다. 부호 뒤에 닫는 따옴표나 괄호가 오면 그 뒤에서 나눈다. `e.g.` 처럼 영문
한 글자 뒤의 마침표는 약어로 본다. 숫자 사이의 마침표는 소수점이라 애초에 공백이 없다.
"""

from __future__ import annotations

import re
from dataclasses import dataclass


@dataclass(frozen=True)
class Sentence:
    text: str
    start: int
    """분석에 넣은 문자열 안의 시작 오프셋."""
    end: int


# 부호 바로 뒤에 공백이 올 때만 나눈다. 닫는 따옴표나 괄호가 끼면 (`"안녕." 하고`) 바깥 문장이 이어지는 것이다.
TERMINAL = re.compile(r"[.?!]+\s+(?=\S)")
ABBREVIATION = re.compile(r"(?:^|[^A-Za-z])[A-Za-z]$")


def splitSentences(text: str) -> list[Sentence]:
    sentences: list[Sentence] = []
    start = 0
    for match in TERMINAL.finditer(text):
        before = text[: match.start()]
        if ABBREVIATION.search(before):
            continue
        end = match.start() + len(match.group(0).rstrip())
        piece = text[start:end].strip()
        if piece:
            sentences.append(Sentence(piece, start + (len(text[start:end]) - len(text[start:end].lstrip())), end))
        start = match.end()
    tail = text[start:].strip()
    if tail:
        lead = len(text[start:]) - len(text[start:].lstrip())
        sentences.append(Sentence(tail, start + lead, start + lead + len(tail)))
    return sentences
